File: app/src/main/java/io/github/shadow578/music_dl/util/notifications/NotificationChannels.java
package io.github.shadow578.music_dl.util.notifications;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.app.NotificationChannelCompat;
import androidx.core.app.NotificationManagerCompat;

/**
 * class to handle notification channels
 */
@SuppressWarnings("unused")
public enum NotificationChannels {
    /**
     * default notification channel.
     * <p>
     * only for use when testing stuff (and the actual channel is not setup yet) or for notifications that are normally not shown
     */
    Default(),

    /**
     * notification channel used by {@link io.github.shadow578.music_dl.downloader.DownloaderService} to show download progress
     */
    DownloadProgress(NotificationManagerCompat.IMPORTANCE_LOW);

// region boring background stuff

    /**
     * prefix for notification channel IDs
     */
    private static final String ID_PREFIX = "io.github.shadow578.youtube_dl.";

    /**
     * display name of this channel.
     */
    @Nullable
    @StringRes
    private final Integer name;

    /**
     * display description of this channel.
     */
    @Nullable
    @StringRes
    private final Integer description;

    /**
     * importance int
     */
    @Nullable
    private final Integer importance;

    /**
     * define a new notification channel. the channel will have no description and a fallback name
     */
    NotificationChannels() {
        name = null;
        description = null;
        importance = null;
    }

    /**
     * define a new notification channel. the channel will have no description and a fallback name
     *
     * @param importance the importance of the channel
     */
    NotificationChannels(int importance) {
        name = null;
        description = null;
        this.importance = importance;
    }

    /**
     * define a new notification channel. the channel will have no description
     *
     * @param name       the display name of the channel
     * @param importance the importance of the channel
     */
    NotificationChannels(@StringRes int name, int importance) {
        this.name = name;
        description = null;
        this.importance = importance;
    }

    /**
     * define a new notification channel
     *
     * @param name        the display name of the channel
     * @param description the display description of the channel
     * @param importance  the importance of the channel
     */
    NotificationChannels(@StringRes int name, @StringRes int description, int importance) {
        this.name = name;
        this.description = description;
        this.importance = importance;
    }

    /**
     * @return id of this channel definition
     */
    @NonNull
    public String id() {
        return ID_PREFIX + name().toUpperCase();
    }

    /**
     * create the notification channel from the definition
     *
     * @param ctx the context to resolve strings in
     * @return the channel, with id, name, desc and importance set
     */
    @NonNull
    private NotificationChannelCompat createChannel(@NonNull Context ctx) {
        // create channel builder
        final int importance;
        if (this.importance != null) {
            importance = this.importance;
        } else {
            importance = NotificationManagerCompat.IMPORTANCE_DEFAULT;
        }
        final NotificationChannelCompat.Builder channelBuilder = new NotificationChannelCompat.Builder(id(), importance);

        // set name with fallback
        channelBuilder.setName(name != null ? ctx.getString(name) : id());

        // set description
        if (description != null) {
            channelBuilder.setDescription(ctx.getString(description));
        }

        return channelBuilder.build();
    }

    /**
     * register all notification channels
     *
     * @param ctx the context to register in
     */
    public static void registerAll(@NonNull Context ctx) {
        // get notification manager
        final NotificationManagerCompat notificationManager = NotificationManagerCompat.from(ctx);

        // register channels
        for (NotificationChannels ch : NotificationChannels.values()) {
            notificationManager.createNotificationChannel(ch.createChannel(ctx));
        }
    }
    //endregion
}

File: app/src/main/kotlin/io/github/shadow578/yodel/util/NotificationChannels.kt
package io.github.shadow578.yodel.util

import android.content.Context
import androidx.annotation.StringRes
import androidx.core.app.*
import io.github.shadow578.yodel.R

/**
 * class to handle notification channels
 *
 * @param displayName display name resource
 * @param description description text resource
 * @param importance importance of this channel
 */
enum class NotificationChannels(
    @StringRes private val displayName: Int? = null,
    @StringRes private val description: Int? = null,
    @StringRes private val importance: Int = NotificationManagerCompat.IMPORTANCE_DEFAULT
) {
    /**
     * default notification channel.
     * <p>
     * only for use when testing stuff (and the actual channel is not setup yet) or for notifications that are normally not shown
     */
    Default(
        R.string.channel_default_name,
        R.string.channel_default_description
    ),

    /**
     * notification channel used by {@link io.github.shadow578.music_dl.downloader.DownloaderService} to show download progress
     */
    DownloadProgress(
        R.string.channel_downloader_name,
        R.string.channel_downloader_description,
        NotificationManagerCompat.IMPORTANCE_LOW
    );


    // region boring background stuff
    /**
     * the id of this channel definition
     */
    val id: String
        get() = "io.github.shadow578.youtube_dl.${name.uppercase()}"

    /**
     * create the notification channel from the definition
     *
     * @param ctx the context to resolve strings in
     * @return the channel, with id, name, desc and importance set
     */
    private fun createChannel(ctx: Context): NotificationChannelCompat {
        return NotificationChannelCompat.Builder(id, importance).apply {
            // set name with fallback
            setName(if (displayName != null) ctx.getString(displayName) else id)

            // set description
            setDescription(if (description != null) ctx.getString(description) else null)
        }.build()
    }

    companion object {
        /**
         * register all notification channels
         *
         * @param ctx the context to register in
         */
        fun registerAll(ctx: Context) {
            // get notification manager
            val notificationManager = NotificationManagerCompat.from(ctx)

            // register channels
            for (ch in values())
                notificationManager.createNotificationChannel(ch.createChannel(ctx))
        }
    }
    //endregion
}